File: nodejs/wgsvr.js
const {exec} = require("child_process");
const dgram = require("dgram");
const udpServer = dgram.createSocket("udp4");
const options = {port: (process.argv[4] || 33055), address: (process.argv[3] || "0.0.0.0")}
const cmd_pre = "wg set " + (process.argv[2] || "wg0") + " peer ";

const workInfos = {};

function removeWork(wis, publicKey){
	for (let key in wis){
		if (wis[key].publicKey === publicKey){
			clearInterval(wis[key].intervalID);
			delete wis[key];
			break;
		}
	}
}

// pingXXX-ip-port reqIP -> pingXXX-ip-port pong
// setxXXX-ip-port publicKey... -> setxXXX-ip-port success
udpServer.on("message", function(msg, rinfo) {
	const msg_s = msg.toString();
	console.log(new Date().getTime(), `udp recv(${rinfo.address}:${rinfo.port}):`, msg_s);
	
	const cmd_id = msg_s.substring(0, msg_s.indexOf(" ") + 1);
	const cmd_data = msg_s.substring(cmd_id.length);
	if (cmd_id.startsWith("ping")){
		const workInfo = workInfos[cmd_data];
		if (workInfo){
			workInfo.tickCount = 0;
		}
		udpServer.send(cmd_id + "pong", rinfo.port,rinfo.address);
	}else{
		// 执行新配置
		exec(cmd_pre + cmd_data, (error, stdout, stderr) => {
			console.log(new Date().getTime(), `stdout: ${stdout}`);
			console.error(new Date().getTime(), `stderr: ${stderr}`);
			if (error) {
				console.error(new Date().getTime(), `exec error: ${error}`);
				udpServer.send(cmd_id + "fail " + error, rinfo.port,rinfo.address);
			}else{
				udpServer.send(cmd_id + "success", rinfo.port,rinfo.address);
				
				const publicKey = cmd_data.substring(0, 44);
				removeWork(workInfos, publicKey);

				// 需要心跳维护
				if (!cmd_data.endsWith("remove") && !cmd_id.startsWith("set0")){
					const pos1 = cmd_data.indexOf("allowed-ips ") + 12;
					const pos2 = cmd_data.indexOf("/", pos1);
					const peerIP = cmd_data.substring(pos1, pos2);					
					
					var workInfo = {publicKey, tickCount:0};
					workInfo.intervalID = setInterval((wi) => {
						//console.log(new Date().getTime(), "setInterval"));
						wi.tickCount ++;
						if (wi.tickCount === 6){ // 30
							removeWork(workInfos, wi.publicKey)
							
							console.log(new Date().getTime(), "remove", wi.publicKey);
							exec(cmd_pre + wi.publicKey + " remove", (error, stdout, stderr) => console.error(`stderr: ${stderr}`));
						}
					}, 5000, workInfo);
					
					workInfos[peerIP] = workInfo;
					//console.log("workInfo", peerIP, workInfo); 
				}
			}
		});
	}
});

udpServer.bind(options);
console.log(new Date().getTime(), `udp server(${options.address}:${options.port}) start...`);

File: lua/wgsvr.lua

local socket = require("socket.core")


-- pingXXX-ip-port reqIP -> pingXXX-ip-port pong
-- setxXXX-ip-port publicKey... -> setxXXX-ip-port success
local function start(device, address, port)
	print(string.format("[%d][wgsvr] open(address=%s, port=%d, device=%s)", os.millisecond(), address, port, device))
	
	local cmd_pre = string.format("wg set %s peer ", device)
	local workInfos = {}
	local m = socket.udp()
	m:settimeout(5) --b,t
	assert(m:setsockname(address, port))
		
	local removeWork = function (wis, publicKey)
		for k,v in pairs(wis) do
			if (v.publicKey == publicKey) then
				wis[k] = nil
				break
			end
		end
	end

	local timeoutWork = function (wis, cb)
		for k,v in pairs(wis) do
			v.tickCount = v.tickCount + 1
			if (v.tickCount == 6) then
				wis[k] = nil
				cb(v)
				break
			end
		end
	end

		
	while(true) do
		local cmd_id,cmd_data
		local data,remoteAddress,remotePort = m:receivefrom()
		if (not data) then
			--print(string.format("[%d][wgsvr] err", os.millisecond()), remoteAddress)			
			timeoutWork(workInfos, function(v)
				print(string.format("[%d][wgsvr] remove(%s)", os.millisecond(), v.publicKey))
				os.execute(cmd_pre .. v.publicKey .. " remove")
			end)
		else
			print(string.format("[%d][wgsvr] receivefrom(%s:%d), data:", os.millisecond(), remoteAddress, remotePort), data)
			cmd_id,cmd_data = data:match("^([^ ]+) (.+)$")
		end
		
		if (cmd_id and cmd_data) then		
			if cmd_id:find("^ping") then
				local workInfo = workInfos[cmd_data]
				if (workInfo) then
					workInfo.tickCount = 0
				end
				m:sendto(cmd_id .. " pong", remoteAddress, remotePort)
			else
				-- 执行新配置
				local ok = os.execute(cmd_pre .. cmd_data)
				assert(m:sendto(cmd_id .. (ok and " success" or " fail"), remoteAddress, remotePort))
				
				if ok then
					local publicKey = cmd_data:sub(1, 1+44)
					removeWork(workInfos, publicKey)

					-- 需要心跳维护
					if (not cmd_data:find("remove$")) and (not cmd_id:find("^set0")) then
						local peerIP = cmd_data:match("allowed%-ips ([^/]+)")
						if peerIP then
							workInfos[peerIP] = {publicKey=publicKey, tickCount=0}
						end
					end
				end
			end
		end
	end

	m:close()

	print(string.format("[%d][wgsvr] close", os.millisecond()))
end

local function daemon(arg1, arg2, arg3)
	while(true) do
		print(string.format("[%d][wgsvr] daemon", os.millisecond()))
		local ok,err = pcall(start, arg1, arg2, arg3)
		if (not ok) then
			print(string.format("[%d][wgsvr] err", os.millisecond))
		end
		os.sleep(15)
	end
end


--arg[1] = "daemon"
if (arg[1] == "daemon") then
	daemon(arg[2], arg[3], arg[4])
else
	start(arg[1] or "wg0", arg[2] or "0.0.0.0", arg[3] or 33055)
end
